### PR TITLE
Move `TrainerCallbackHookMixin.on_save/load_checkpoint`  to `Trainer` and rename for clarity

### LINKED ISSUE #11165: Move `TrainerCallbackHookMixin.on_save/load_checkpoint`  to `CheckpointConnector` and rename for clarity
_Originally posted by @ananthsub in https://github.com/PyTorchLightning/pytorch-lightning/pull/11148#discussion_r771886148_

## Proposed refactor
* Move `TrainerCallbackHookMixin.on_save/load_checkpoint` to `CheckpointConnector` as functions
* Rename them to something like `call_callbacks_on_save_checkpoint` and `call_callbacks_on_load_checkpoint`

### Motivation

* We are planning to deprecate `TrainerCallbackHookMixin` in #11148. However there are still two methods in there that we need to keep: `on_save_checkpoint` and `on_load_checkpoint`
* `CheckpointConnector` is a more logical place for these functions, since that is where all of the checkpointing code lives.
* Also, the current naming is confusing because it doesn't convey that we are calling the hook on all of the callbacks

### Pitch
* Move the functions to `CheckpointConnector` and rename them to `call_callbacks_on_save_checkpoint` and `call_callbacks_on_load_checkpoint`
* Deprecate the original methods in `TrainerCallbackHookMixin` as part of #11148

cc @justusschock @awaelchli @akihironitta @ananthsub @ninginthecloud @daniellepintz

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?
Fixes #11165

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2287,6 +2289,47 @@ def terminate_on_nan(self, val: bool) -> None:
         )
         self._terminate_on_nan = val  # : 212
 
+    def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[str, dict]:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2292]: move this to where `_call_callback_hooks` is defined so they're next to each other. in a comment, could you describe why this is a dedicated function instead of being part of `_call_callback_hooks` ?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2287,6 +2289,47 @@ def terminate_on_nan(self, val: bool) -> None:
...
+        """Called when saving a model checkpoint.
+
+        Calls every callback's `on_save_checkpoint` hook.
+        """
+        callback_states = {}
+        for callback in self.callbacks:
+            state = callback.on_save_checkpoint(self, self.lightning_module, checkpoint)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2299]: a follow up should be to add profiling for each of these calls

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2292]: @jjenniferdai - this is where having a separate `state_dict/load_state_dict` to collect and load states from individual components vs `on_save/load_checkpoint` will be useful. one operates on the local state (per component), one operates on the global checkpoint (across all components)

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2287,6 +2289,47 @@ def terminate_on_nan(self, val: bool) -> None:
...
+        return callback_states
+
+    def _call_callbacks_on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """Called when loading a model checkpoint.
+
+        Calls every callback's `on_load_checkpoint` hook.
+        """
+        # Todo: the `callback_states` are dropped with TPUSpawn as they
+        # can't be saved using `xm.save`
+        # https://github.com/pytorch/xla/issues/2773
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2311]: @kaushikb11 is this comment here relevant? i don't see any special handling for this here. 

I see it's also referenced here, where the callback states are explicitly not saved: https://github.com/PyTorchLightning/pytorch-lightning/blob/08e661ff7266f298a6e904fa04931698e20b2b89/pytorch_lightning/plugins/io/xla_plugin.py#L43-L47

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -2287,6 +2289,47 @@ def terminate_on_nan(self, val: bool) -> None:
...
+                "Be aware that when using `ckpt_path`,"
+                " callbacks used to create the checkpoint need to be provided during `Trainer` instantiation."
+                f" Please add the following callbacks: {list(difference)}.",
+            )
+
+        for callback in self.callbacks:
+            state = callback_states.get(callback.state_key, callback_states.get(callback._legacy_state_key))
+            if state:
+                state = deepcopy(state)
+                callback.on_load_checkpoint(self, self.lightning_module, state)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2331]: same here for profiling

**Comment by awaelchli** [comment]
Was the intention to deprecate the definitions TrainerCallbackHookMixin ?

**Comment by daniellepintz** [comment]
@awaelchli sorry for the confusion, the deprecation of the entire `TrainerCallbackHookMixin` is being done in https://github.com/PyTorchLightning/pytorch-lightning/pull/11148

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1566,6 +1568,51 @@ def _call_callback_hooks(
...
+        return callback_states
+
+    def _call_callbacks_on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """Called when loading a model checkpoint.
+
+        Calls every callback's `on_load_checkpoint` hook.
+        """
+        # Todo: the `callback_states` are dropped with TPUSpawn as they
+        # can't be saved using `xm.save`
+        # https://github.com/pytorch/xla/issues/2773
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1593]:
```suggestion
```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -1566,6 +1568,51 @@ def _call_callback_hooks(
             # restore current_fx when nested context
             pl_module._current_fx_name = prev_fx_name
 
+    # dedicated functions for calling callbacks' `on_save_checkpoint` and `on_load_checkpoint`
+    # hooks to include special logic for storing of callback_states
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1572]: let's integrate this into the docstring(s)?

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -1566,6 +1568,51 @@ def _call_callback_hooks(
...
+    def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[str, dict]:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1573]: let's integrate this into the docstring(s)?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py]: planning to file this issue this week!

**Review by Borda — APPROVED** [review]

**Comment by daniellepintz** [comment]
Ready to be merged!
